### PR TITLE
add varargs nondiff rules

### PR DESCRIPTION
**Pull request by niklasschmitz**
Addresses #253, following @oxinabox.

I left out the last three rand definitions in light of #263 (?)
```julia
@non_differentiable rand(::Any...)
@non_differentiable randn(::Any...)
@non_differentiable randexp(::Any...)
```

### CODE REVIEW

**Review by mzgubic — COMMENTED** [review]
Hey, thanks for taking this on, we appreciate it! The changes look great, thanks for going through the all rules.

With regard to #263: The issue was that the `rand` function arguments were not constrained at all (i.e. were `Any`), meaning that the `@non_differentiable` macro would apply to all the methods, even those that were differentiable in some arguments. It is still possible to reduce the number of lines by replacing, for example:
```
@non_differentiable rand(::AbstractRNG, ::Integer)
@non_differentiable rand(::AbstractRNG, ::Integer, ::Integer)
@non_differentiable rand(::AbstractRNG, ::Integer, ::Integer, ::Integer)
@non_differentiable rand(::AbstractRNG, ::Integer, ::Integer, ::Integer, ::Integer)
@non_differentiable rand(::AbstractRNG, ::Integer, ::Integer, ::Integer, ::Integer, ::Integer)
```
with
```
@non_differentiable rand(::AbstractRNG, ::Integer...)
```

If you can bump the version number as well this PR is ready to merge.

File: src/rulesets/Random/random.jl [hunk 1]
```diff
@@ -15,51 +15,27 @@ end
...
-@non_differentiable rand(::Integer)
-@non_differentiable rand(::Integer, ::Integer)
-@non_differentiable rand(::Integer, ::Integer, ::Integer)
-@non_differentiable rand(::Integer, ::Integer, ::Integer, ::Integer)
-@non_differentiable rand(::Integer, ::Integer, ::Integer, ::Integer, ::Integer)
-
+@non_differentiable rand(::Type{<:Real}, ::Integer...)
+@non_differentiable rand(::Integer...)
 
 # There are many different 1-3 arg methods, but not varargs
```
Comment (niklasschmitz) [inline src/rulesets/Random/random.jl:24]: I'm unsure about what this comment means, so I didn't replace it by `rand!(::Any...)`

File: src/rulesets/Random/random.jl [hunk 2]
```diff
@@ -15,51 +15,27 @@ end
 @non_differentiable rand()
 @non_differentiable rand(::AbstractRNG)
```
Comment (mzgubic) [inline src/rulesets/Random/random.jl:16]: as can this one

**Review by mzgubic — COMMENTED** [review]
Great, thanks! Just a couple of minor things left

File: src/rulesets/Random/random.jl [hunk 3]
```diff
@@ -15,51 +15,27 @@ end
...
-@non_differentiable randn(::Any, ::Any, ::Any)
-@non_differentiable randn(::Any, ::Any, ::Any, ::Any)
-@non_differentiable randn(::Any, ::Any, ::Any, ::Any, ::Any)
+@non_differentiable randn(::Any...)
 
 @non_differentiable randn!(::AbstractArray)
 @non_differentiable randn!(::AbstractRNG, ::AbstractArray)
 
-
-@non_differentiable randn(::AbstractRNG)
```
[the remark this answers is not in the document]
Comment (niklasschmitz) [reply src/rulesets/Random/random.jl:62]: I think it is covered already by `randn(::Any...)` above

File: src/rulesets/Random/random.jl (hunk 1 above)
Comment (niklasschmitz) [reply src/rulesets/Random/random.jl:22]: Now I see, thank you!

**Comment by niklasschmitz** [comment]
Thanks @mzgubic !

**Pull request merged**